airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/record.py

```diff
@@ -0,0 +1,119 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+
+from io import TextIOWrapper
+from json import loads
+from typing import Any, Iterable, List, Mapping, Optional, Union
+
+from .query import ShopifyBulkQuery
+from .tools import END_OF_FILE, BulkTools
+
+
+class ShopifyBulkRecord:
+    def __init__(self, query: ShopifyBulkQuery) -> None:
+        self.composition = query.record_composition
+        self.record_process_components = query.record_process_components
+        self.components: List[str] = self.composition.get("record_components", []) if self.composition else []
```

Comment:
`Record Component` is a part of the record, that has been fetched.
Example:

stream: `Discount_Codes`

When we asking  for the data for this  stream,  the Shopify collects  it like:
```jsonl
{"__typename": "DiscountCode", "id": "some_parent_id_here"},
{"__typename": "DiscountReedemCode", "id": "some_child_id_here"},
```
This should be a single record in the output, so the line with `{"__typename": "DiscountCode"}`  - main record
and `{"__typename": "DiscountReedemCode"}` is a component of the main record that should be joined back to the main record in a way defined in `query.record_process_components()` method.

This would be described in a Playbook as well.